### DISCUSSION TITLE
Need strategy help for this kind of statement

### DISCUSSION THREAD

**Discussion by Hubro** [tree-sitter/tree-sitter#1269] [Q&A] [answered]
I am trying to write a Tree-sitter parser for YANG, which has a pretty simple grammar.

Put very simply, the language can be thought of as having tree kinds of statements:

- A keyword plus an argument, for example `type string`
- A keyword, an argument and a block, for example:

    ```yang
    container foo {
      type string;
      // ...
    }
    ```
- A keyword and a block that contains more statements (no argument), for example:

    ```yang
    input {
      // ...
    }
    ```

To me, it seemed intuitive to write this grammar as:

```javascript
statement: $ => seq(
    choice(
        $.statement_keyword,
        $.extension_keyword,
    ),
    // A statement can either have
    // - An argument
    // - An argument and a block
    // - Just a block
    choice(
        // An argument
        seq($._sep, $.argument, ';'),

        // An argument and a block
        seq($._sep, $.argument, $.block),

        // Just a block
        $.block
    )
),
```

This works beautifully for the for the first two kinds of statements, but gives errrors for the third kind:

```yang
module my-module-name {
    augment /ncs:devices/device {
        container new-container {
            tailf:action add-device {
                tailf:actionpoint my-module-name-add-device;
                tailf:info
                    "Add a device!";
                input {
                    uses inventory-actions:device-input;
                }
                output {
                    leaf result {
                        type string;
                        description "Action result.";
                    }
                }
            }
        }
    }
}
```

![image](https://user-images.githubusercontent.com/597206/125728491-a979d349-3993-4f4c-9798-97b2a2432aa8.png)

(Used a screenshot so you see the colors)

Both the input and output blocks here have an "(ERROR)" in the syntax tree. This makes me think that Tree-sitter went down the "keyword + value + block" route and simply inserted an ERROR node in place of the value, because the value is missing. But why does Tree-sitter go down this route when I've placed a valid alternative in the same `choice` statement? Is this just a limitation of Tree-sitter? What can I do instead?

---

[Full code here](https://github.com/Hubro/tree-sitter-yang/blob/HELP/grammar.js#L46)

**Comment by ahlinc** [comment, marked as answer]
Why do you want to have the explicit `$._sep` token?
If to try to rewrite the `statement` rule like:
```js
        statement: $ => seq(
            choice(
                $.statement_keyword,
                $.extension_keyword,
            ),
            choice(
                seq($.argument, choice($.block, ';')),
                $.block
            )
        ),
```
It seems to work fine with your example.

**Reply by Hubro** [reply]
It was my understanding that Tree-sitter implicitly allows zero or more whitespace between each token. Doesn't that mean that if I don't explicitly specify that one or more spaces is required between the keyword and argument, this becomes legal?

```yang
typestring;
```

But it does make sense that it works now, I guess the `extras` patterns already gobble up the whitespace before my patterns get a chance, so the match fails? Or something like that :thinking:

**Reply by Hubro** [reply]
Huh, apparently not... Now I'm utterly confused. If I write `seq($.foo, $.bar)`, where `foo: $ => 'foo'` and `bar: $ => 'bar'`, I would expect it to successfully match "foobar", but it doesn't :thinking: It requires a space between "foo" and "bar".

But then how would I match `foobar` as two tokens if I wanted to? :thinking:

**Reply by ahlinc** [reply]
> It was my understanding that Tree-sitter implicitly allows zero or more whitespace between each token.

Correct.
And you can use `' '` to require a one space between tokens (but this will make accessible anonymous node in your grammar.
To avoid it to be in the tree you can use it through underscore prefixed name like `_sep: $ => ' '` or use a regexp '/ /' or `/\s/` right in `seq(..., /\s/, ...)` between tokens cause all unnamed regex nodes are absent in the resulting tree but correctly checked during the parsing process.

Also note: 
- ~That writing regex like `/\s+/` doesn't have a lot of sense cause it's no matter what will consume more spaces this regex or a regex in the `extras` if you have similar one in it.~
- In the `extras` it also have no sense to write `/\s+/`, writing `/\s/` is enough cause the extras are repeatable things and for the `extras: $ => [/\s/]` it's like `repeat(/\s/)` that are allowed everywhere between other tokens.

**Edit (week after):**
Above my second note is _correct_ but the first was wrong. If you have some token that that compete with the `extras`, spaces definition it's important to define in the way that it will consume all related input once so there will be only one terminal token, so for spaces it should be `/\s+/` the regexp with repetition. **Otherwise such token definition like `/\s/` will produce series of space tokens of arbitrary size that would break both `LR(1)` and `GLR` algorithms logic.**

> If I write `seq($.foo, $.bar)`, where `foo: $ => 'foo'` and `bar: $ => 'bar'`, I would expect it to successfully match "foobar", but it doesn't  It requires a space between "foo" and "bar".

I don't understand how you achieved that for such simple case, `foobar` and `foo bar` should be parsed to the same tree and it's parsed correctly on a test I've made.

**Reply by ahlinc** [reply]
If you need that `$._sep` in your `statement` rule, you can write your rule like:
```js
        statement: $ => seq(
            choice(
                $.statement_keyword,
                $.extension_keyword,
            ),
            $._sep,
            choice(
                seq($.argument, choice($.block, ';')),
                $.block
            )
        ),
```
It's not fully semantically equal to what you wrote but this also will work.

**Update:**
With your initial version of the `statement` rule you've got in interesting situation and it doesn't work:
- Cause in your `choice` you required `$._sep` and it parsed correctly in the `_sep` token. You can verify this with the `-d` flag for the `tree-sitter parse` command.
- But after the `$._sep` there is a `{` token that is a beginning token for the `$.block` rule. And the parser can't ignore that it already has parsed `$._sep` token and just switch to another the `$.block` branch in the `choice` statement.
- That's the reason why the parser firstly decides to generate the `ERROR` node for `$._sep` token and after that do the switch to the `$.block` branch in the `choice` statement.

To fix the above described situation you can rewrite your rule like the following:
```js
        statement: $ => seq(
            choice(
                $.statement_keyword,
                $.extension_keyword,
            ),
            choice(
                seq($._sep, $.argument, ';'),
                seq($._sep, $.argument, $.block),
                seq(optional($._sep), $.block),
            )
        ),
```
Which is also works fine and means exactly what you wrote. All that you were needed is to explicitly say that the `$._sep` token can be parsed or skipped for the `$.block` branch.

**Reply by Hubro** [reply]
> Also note that writing regex like /\s+/ doesn't have a lot of sense cause it's no matter what will consume more spaces this regex or a regex in the extras if you have similar one in it. In the extras it also have no sense to write /\s+/, writing /\s/ is enough cause the extras are repeatable things and for the extras: $ => [/\s/] it's like repeat(/\s/) that are allowed everywhere between other tokens.

Ah, that makes sense, thanks!

> I don't understand how you achieved that for such simple case, foobar and foo bar should be parsed to the same tree and it's parsed correctly on a test I've made.

I quickly tested it in my current grammar, but I must have screwed up somehow... I tested it again now, and now it accepts both `foo bar` and `foobar`... :confused: 

> Which is also works fine and means exactly what you wrote. All that you were needed is to explicitly say that the $._sep token can be parsed or skipped for the $.block branch.

Aah, nice! That's a perfect fix. It's really unintuitive how `extras` affects rules, but I think I'll  get the hang of it.
